---
title: "Permuted phenotypes 3"
author: "Alice MacQueen"
date: "12/2/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(bigsnpr)
# library(switchgrassGWAS)

# library(snpdiver)

```

# GxE Group suggestions

1. Just permute sample.ID relative to all the phenotypes. This will keep the correlations between phenotypes within individuals. Then you might recover these correlations still within mash.

2. Add a constant to a trait value for just one subpopulation before GWAS. Can you recover this spiked in signal in GWAS? Or does your population structure correction control for it?

3. More complex analysis: make a covariance matrix with just ancestry informative markers for your phenotype. What does this matrix look like?

    * Also keep and compare MAF 5% and MAF 10% GWAS. And investigate if I can just subset GWAS results to MAF 10% results. That would be better in the long run. I don't see why that wouldn't be possible given the snp_maf function and the subset & big_copy functions.

    * Want to write an article for switchgrassGWAS also, on how to make a table & specify the exact number of PCs to use, in case you like this correction but don't believe in whatever lambda_GC this function is programmed to find. Also, mention that there's GAPIT but that it will take a lot longer. 

## Set up permuted data with sample
```{r}
metadata <- readRDS(here("data/metadata.rds"))
prefix <- "pvdiv_panicles_2019_BLUPs_PLANT_ID_geno_subset_"

phe_gwas <- read_csv(here("data", 
                          paste0(prefix, "BLUP_phenotypes",
                                              ".csv"))) %>%
  left_join(select(metadata, PLANT_ID, SUBPOP, ECOTYPE_NNET)) %>%
  select(PLANT_ID, SUBPOP, ECOTYPE_NNET, everything()) %>% 
  pivot_longer(4:15, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE)) 
```

```{r}
# permutations of PLANT_ID across all phenotyped individuals
phe_perm_pid1 <- phe_gwas %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, 
              names_prefix = "PERM_PID_1_") %>%
  ungroup() %>%
  mutate(sample.ID = sample(PLANT_ID, replace = FALSE)) %>%
  select(PLANT_ID, sample.ID, everything()) %>%
  select(-SUBPOP, -ECOTYPE_NNET, -PLANT_ID) %>%
  rename(PLANT_ID = sample.ID)
phe_perm_pid2 <- phe_gwas %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, 
              names_prefix = "PERM_PID_2_") %>%
  ungroup() %>%
  mutate(sample.ID = sample(PLANT_ID, replace = FALSE)) %>%
  select(PLANT_ID, sample.ID, everything()) %>%
  select(-SUBPOP, -ECOTYPE_NNET, -PLANT_ID) %>%
  rename(PLANT_ID = sample.ID)

# permutations of PLANT_ID within subpopulations
phe_perm_pid_sub <- phe_gwas %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, 
              names_prefix = "PID_PERM_SUB_") %>%
  group_by(SUBPOP) %>%
  mutate(sample.ID = sample(PLANT_ID, replace = FALSE)) %>%
  select(PLANT_ID, sample.ID, everything()) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET, -PLANT_ID) %>%
  rename(PLANT_ID = sample.ID)

# permutations of PLANT_ID within subpopulations & ecotypes
phe_perm_pid_sub_eco <- phe_gwas %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, 
              names_prefix = "PID_PERM_SUB_ECO_") %>%
  group_by(SUBPOP, ECOTYPE_NNET) %>%
  mutate(sample.ID = sample(PLANT_ID, replace = FALSE)) %>%
  select(PLANT_ID, sample.ID, everything()) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET, -PLANT_ID) %>%
  rename(PLANT_ID = sample.ID)

# permutations of PLANT_ID within collection sites
phe_perm_pid_site <- phe_gwas %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, 
              names_prefix = "PID_PERM_SITE_") %>%
  separate(PLANT_ID, into = c("COL_SITE", "PLANT_ID_IND"), sep = -1, 
         remove = FALSE) %>%
  group_by(COL_SITE) %>%
  mutate(sample.ID = sample(PLANT_ID, replace = FALSE)) %>%
  select(PLANT_ID, sample.ID, everything()) %>%
  ungroup() %>%
  select(-COL_SITE, -PLANT_ID_IND, -SUBPOP, -ECOTYPE_NNET, -PLANT_ID) %>%
  rename(PLANT_ID = sample.ID)

 
```

Join these permuted PLANT_ID dataframes together
```{r}
phe_blups <- phe_gwas %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, names_prefix = "BLUP_") %>%
  select(-SUBPOP, -ECOTYPE_NNET)

(all_raw_reps <- readRDS(file = here("data/all_replicates_raw_phenotypes_geno_subset.rds")) %>%
  rename(PLANT_ID = sample.ID))

all_phe <- phe_blups %>%
  full_join(all_raw_reps, by = "PLANT_ID") %>%
  full_join(phe_perm_pid_site, by = "PLANT_ID") %>%
  full_join(phe_perm_pid_sub_eco, by = "PLANT_ID") %>%
  full_join(phe_perm_pid_sub, by = "PLANT_ID") %>%
  full_join(phe_perm_pid1, by = "PLANT_ID") %>%
  full_join(phe_perm_pid2, by = "PLANT_ID") 
  #select(PLANT_ID, contains("PL_PKLE"), contains("PL_CLMB"), 
  #       contains("PL_KBSM"), contains("PBN_PKLE"),
  #       contains("PBN_CLMB"), contains("PBN_KBSM"), 
  #       contains("SBN_PKLE"),
  #       contains("SBN_CLMB"), contains("SBN_KBSM"), everything()
  #       )
saveRDS(all_phe, file = here("data", "all_panicle_phe_perm_pid.rds"))

```


# GWAS

```{r}
library(tidyverse)
library(here)
library(bigsnpr)
library(switchgrassGWAS)

snp05 <- snp_attach("~/Github/PanicleData/9-DiversityPanel/data/Pvirgatum_V5_GWAS_381g_PanicleData_subset_maf_0.05.rds")
covar05 <- readRDS(here("analysis", "gwas", "PID_PERM", "SVD_381g_12.8M_SNPs_15PCs.rds"))

all_phe <- readRDS(file = here("data", "all_panicle_phe_perm_pid.rds"))

if (!dir.exists(here("analysis", "gwas", "PID_PERM2"))) {
  dir.create(here("analysis", "gwas", "PID_PERM2"))
}
pvdiv_standard_gwas(snp = snp05, df = all_phe, covar = covar05, 
                    type = "linear", outputdir = here("analysis", "gwas",
                                                      "PID_PERM2"), 
                    saveplots = FALSE, savegwas = TRUE, savetype = 'fbm', 
                    suffix = "all_panicle_phe_perm_pid")

```

These are the GWAS effects for the second kind of permuted dataset:
```{r}
library(snpdiver)

snp05 <- snp_attach("~/Github/PanicleData/9-DiversityPanel/data/Pvirgatum_V5_GWAS_381g_PanicleData_subset_maf_0.05.rds")
  
effectspid <- big_attach(here("analysis/gwas/PID_PERM/gwas_effects_all_panicle_phe_perm_pid.rds"))
metadatapid <- read_csv(here("analysis/gwas/PID_PERM/gwas_effects_all_panicle_phe_perm_pid_associated_metadata.csv"))

mtest5 <- dive_effects2mash(effects = effectspid, snp = snp05, 
                           metadata = metadatapid, 
                  suffix = "test_mash_5", 
                  outputdir = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/PID_PERM")

saveRDS(mtest5, file = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/PID_PERM/Mash_output_all_panicle_phe_perm_pid.rds" )
        

```


2. Add a constant to a trait value for just one subpopulation before GWAS. Can you recover this spiked in signal in GWAS? Or does your population structure correction control for it?

3. More complex analysis: make a covariance matrix with just ancestry informative markers for your phenotype. What does this matrix look like?



# -----------

# Mash results on two types of permuted dataset

This is the SNP file used for both types of permutation analysis.
```{r}
library(tidyverse)
library(bigsnpr)
load_all("~/Github/snpdiver/")

snp05 <- snp_attach("~/Github/PanicleData/9-DiversityPanel/data/Pvirgatum_V5_GWAS_381g_PanicleData_subset_maf_0.05.rds")
metadata <- readRDS("~/Github/PanicleData/9-DiversityPanel/data/metadata.rds")
subpops <- readRDS("~/Github/pvdiv-phenology-gxe/data/subpop_color_coding.rds")
pid_phe <- readRDS(file = file.path("~/Github/PanicleData/9-DiversityPanel/",
                                    "data", "all_panicle_phe_perm_pid.rds"))

```


These are the GWAS effects for the first kind of permuted dataset
```{r}
effects_perm1 <- big_attach(here("analysis/gwas/MAF05/gwas_effects_all_panicle_phe_MAF_5per.rds"))
(metadata_perm1 <- read_csv(here("analysis/gwas/MAF05/gwas_effects_all_panicle_phe_MAF_5per_associated_metadata.csv")))

metadata_perm1$phe[c(1:3, 12:14, 23:25, 34:36, 45:47, 56:58, 67:69, 78:80,
                     89:91)]
BLUP_PERM_1 <- c(1:3, 12:14, 23:25, 34:36, 45:47, 56:58, 67:69, 78:80,
                     89:91)
PERM_1 <- c(2:3, 13:14, 24:25, 35:36, 46:47, 57:58, 68:69, 79:80,
                     90:91)
```

These are the GWAS effects for the second kind of permuted dataset:
```{r}
effects_pid <- big_attach(here("analysis/gwas/PID_PERM/gwas_effects_all_panicle_phe_perm_pid.rds"))
(metadata_pid <- read_csv(here("analysis/gwas/PID_PERM/gwas_effects_all_panicle_phe_perm_pid_associated_metadata.csv")))

metadata_pid$phe[c(1:9, 82:99)]
BLUP_PERM_2 <- c(1:9, 82:99)
PERM_2 <- c(82:99)
```


# Subset to speed mash
Subset SNP & effects
```{r}
# Set up SNP subset given the PLANT_ID to include, the maf, and reduce to a 
# small set of SNPs to test
plants <- snp05$fam$sample.ID
panicle_381 <- which(plants %in% pid_phe$PLANT_ID)
maf_thresh10 <- snp_MAF(snp05$genotypes, ind.row = panicle_381)
snp_subset_maf10 <- which(maf_thresh10 > 0.1)
snp_subset_maf10_10k <- c(1:floor(length(snp_subset_maf10)/1000))*1000
snp_subset <- snp_subset_maf10[snp_subset_maf10_10k]

# subset SNPs & effects given the above snp_subset of snp05
snp05_sub <- subset(snp05, ind.col = snp_subset)
snp_05s <- snp_attach(snp05_sub)

```

Subset Perm1 & run mash on BLUPs & PERM
```{r}
effects_sub1 <- big_copy(effects_perm1, ind.row = snp_subset)

# pick the phenotype subset

mtest1 <- dive_effects2mash(effects = effects_sub1, snp = snp_05s, 
                            scale.phe = FALSE, replace.na = FALSE,  
                            metadata = metadata_perm1, phe = BLUP_PERM_1,
                            suffix = "BLUP_PERM_1_maf10_10ksubset", 
                            outputdir = here("analysis/gwas/MAF05"))
saveRDS(mtest1, file = here("analysis/gwas/MAF05/", 
                            paste0("Mash_output_", "1K_strong_", "2K_random_",
                                   "BLUP_PERM_1_maf10_10ksubset", ".rds" )))

```


Subset PID & run mash on BLUPs & PERM
```{r}
effects_sub2 <- big_copy(effects_pid, ind.row = snp_subset)

# pick the phenotype subset (BLUPs & total PERM)

mtest2 <- dive_effects2mash(effects = effects_sub2, snp = snp_05s, 
                            scale.phe = FALSE, replace.na = FALSE,  
                            metadata = metadata_pid, phe = BLUP_PERM_2,
                            suffix = "BLUP_PERM_2_maf10_10ksubset", 
                            outputdir = here("analysis/gwas/PID_PERM"))

saveRDS(mtest2, file = here("analysis/gwas/PID_PERM/", 
                            paste0("Mash_output_", "1K_strong_", "2K_random_",
                                   "BLUP_PERM_2_maf10_10ksubset", ".rds")))
```


Quite clearly in the 1:9, 55:99 (PERM & BLUP) - the equal_effects covariance structure is strongly in every condition. So when phenotypes are not correlated within genotypes, this structure goes away (as in the first set of permutations I did). But when they are correlated within genotypes, this structure is present & common.



```{r}
mash_plot_covar(mtest1, saveoutput = TRUE)
mash_plot_covar(mtest2, saveoutput = TRUE)
(tpca <- mash_plot_Ulist(mtest1, range = 6, labels = FALSE))
mash_plot_Ulist(mtest2, range = 6)
length(get_significant_results(mtest1))
length(get_significant_results(mtest2))
mash_plot_manhattan_by_condition(mtest1, snp = snp_05s, saveoutput = TRUE)
mash_plot_marker_effect(mtest1, snp = snp_05s, n = 1, saveoutput = TRUE)
mash_plot_marker_effect(mtest2, snp = snp_05s, n = 1, saveoutput = TRUE)

mash_plot_sig_by_condition(mtest1, which(grepl("PERM", switchgrassGWAS:::get_colnames(mtest1)) == TRUE), saveoutput = TRUE)

snpdiver::mash_plot_Ulist(mtest1, range = 1:6, labels = FALSE, saveoutput = TRUE)

```