---
title: "Analysis permuted phenotypes"
author: "Alice MacQueen"
date: "11/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(bigsnpr)
library(snpdiver)

```

set up 2-3 panicle data mash runs: completely permuted phenotypes, phenotypes permuted within subpopulations/ecotypes, GWAS on each replicate & compare their overlaps, sample from these replicates for other GWAS... what are major conserved hits if any? 

Here are my notes I have from the last meeting with Tom:
shuffle phenotypes & genotypes, make a null with no genotype-phenotype association, and run mash, look at what you find
try and rerun mash on just the subpopulations, do the loadings on the covariance matrices change?


# Mash with permuted BLUPs

## Set up permuted data with sample
```{r}
metadata <- readRDS(here("data/metadata.rds"))
prefix <- "pvdiv_panicles_2019_BLUPs_PLANT_ID_geno_subset_"

phe_gwas <- read_csv(here("data", 
                          paste0(prefix, "BLUP_phenotypes",
                                              ".csv"))) %>%
  left_join(select(metadata, PLANT_ID, SUBPOP, ECOTYPE_NNET)) %>%
  select(PLANT_ID, SUBPOP, ECOTYPE_NNET, everything())

phe_gwas1 %>%
  filter(SUBPOP != "4X") %>%
  group_by(ECOTYPE, SUBPOP) %>%
  mutate(SBN_KBSM_1 = sample(SBN_KBSM),
         PL_PKLE_1 = sample(PL_PKLE),
         PBN_PKLE_1 = sample(PBN_PKLE)
         ) %>%
  arrange(ECOTYPE, SUBPOP) # convince myself this sampling is obeying group_by

# Then resample for all phenotypes in three different sets
permuted_phe_subpop <- phe_gwas %>% 
  pivot_longer(4:15, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE)) %>%
  group_by(PHE_SITE, SUBPOP) %>%
  mutate(MEAS = sample(MEAS)) %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, names_prefix = "SUB_PERM_")
permuted_phe_subpop_ecotype <- phe_gwas %>% 
  pivot_longer(4:15, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE) & SUBPOP != "4X") %>%
  group_by(PHE_SITE, SUBPOP, ECOTYPE_NNET) %>%
  mutate(MEAS = sample(MEAS)) %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, names_prefix = "SUB_ECO_PERM_")
permuted_phe <- phe_gwas %>% 
  pivot_longer(4:15, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE)) %>%
  group_by(PHE_SITE) %>%
  mutate(MEAS = sample(MEAS)) %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, names_prefix = "ALL_PERM_")

phe_notperm1 <- phe_gwas %>% 
  pivot_longer(4:15, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE)) %>%
  separate(PLANT_ID, into = c("COL_SITE", "PLANT_ID_IND"), sep = -1, remove = FALSE) %>%
  group_by(COL_SITE) %>% tally() %>% arrange(n) %>%
  filter(n <= 9)
phe_notperm2 <- phe_gwas %>%
  pivot_longer(4:15, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE)) %>%
  separate(PLANT_ID, into = c("COL_SITE", "PLANT_ID_IND"), sep = -1, remove = FALSE) %>%
  filter(COL_SITE %in% phe_notperm1$COL_SITE)

ind_perm_phe <- phe_gwas %>% 
  pivot_longer(4:15, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE)) %>%
  separate(PLANT_ID, into = c("COL_SITE", "PLANT_ID_IND"), sep = -1, remove = FALSE) %>%
  group_by(COL_SITE) %>%
  filter(!COL_SITE %in% phe_notperm1$COL_SITE) %>%
  group_by(COL_SITE, PHE_SITE) %>%
  mutate(MEAS = sample(MEAS)) %>%
  full_join(phe_notperm2) %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, names_prefix = "IND_PERM_") %>%
  ungroup() %>%
  select(-COL_SITE, -PLANT_ID_IND, -SUBPOP, -ECOTYPE_NNET) # 271 of 381 individuals permuted within collection sites

saveRDS(permuted_phe, here("data/permuted_phenotypes_geno_subset.rds"))
saveRDS(permuted_phe_subpop, here("data/permuted_phenotypes_within_subpops_geno_subset.rds"))
saveRDS(permuted_phe_subpop_ecotype, here("data/permuted_phenotypes_within_subpop_and_ecotype_geno_subset.rds"))

all_permuted_phe <- permuted_phe %>%
  full_join(permuted_phe_subpop) %>%
  full_join(permuted_phe_subpop_ecotype) %>%
  full_join(ind_perm_phe) %>%
  filter(!is.na(PLANT_ID))
colSums(!is.na(all_permuted_phe))

saveRDS(all_permuted_phe, here("data/all_permuted_phenotypes_geno_subset.rds"))
```

## Set up GWAS & mash on all_panicle phe


```{r}
(all_permuted_phe <- readRDS(here("data/all_permuted_phenotypes_geno_subset.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID))

subpop_v2 <- list(Gulf_and_Midwest = c("Gulf", "Midwest"), Midwest = "Midwest", 
                  Gulf = "Gulf", Atlantic = "Atlantic", 
                  Atlantic_and_Midwest = c("Atlantic", "Midwest"),
                  Atlantic_and_Gulf = c("Atlantic", "Gulf"), 
                  All = c("Atlantic", "Gulf", "Midwest", "4X"))
k = 7
suffix_svd <- paste0("big_three_garden_SNPs", names(subpop_v2)[k])
pavir_svd <- readRDS(file = here("analysis", "mash", paste0(suffix_svd,
                                                        "_svd.rds")))
  
bigsnp_inputs <- read_delim(here("analysis", 
                                   "bigsnp_inputs.txt"),  delim = " ", 
                                   col_names = "SNPfiles") 

```


## Run dive_phe2effects
```{r}
if (!dir.exists(here("analysis", "mash", "permutations"))) {
  dir.create(here("analysis", "mash", "permutations"))
}

suffix = "permuted_panicle_phe"
pavir_snp <- snp_attach(bigsnp_inputs$SNPfiles[k])

m_pan <- dive_phe2effects(df = all_permuted_phe, snp = pavir_snp, svd = pavir_svd,
                       type = "linear", suffix = suffix, ncores = 12, 
                       outputdir = here("analysis", "mash", "permutations"))


m_pan2 <- dive_effects2mash(effects = m_pan$effects, snp = pavir_snp, 
                            metadata = m_pan$metadata, 
                            suffix = "permuted_panicle_phe_5Kstrong_100Krandom",
                            outputdir = here("analysis", "mash",
                                             "permutations"), ncores = 12,
                            num.strong = 5000, num.random = 100000)

  saveRDS(m_pan2, file = here("analysis", "mash", "permutations", 
                             paste0("Full_mash_model_",
                                    "5000", "_SNPs_U_ed_and_",
                                    "100000", "_SNPs_mash_fit_",
                                                    suffix, ".rds")))
  mash_plot_covar(m_pan2, saveoutput = TRUE, suffix = suffix)

  manhattan <- mash_plot_manhattan_by_condition(m_pan2, snp = pavir_snp, 
                                                saveoutput = TRUE)
```

## Upset plots of permutations

### functions to make data frame for single SNPs for Upset plots
```{r}

library(ComplexUpset)
library(cowplot)
library(matrixStats)

thresholdFBM <- function(X, ind, thr) { #, quantile = NA) {
  #if(!is.na(quantile)) {
  #thr <- quantile(X[, ind], quantile)
  #}
  case_when(X[, ind] > thr ~ TRUE,
            TRUE ~ FALSE)
}

big_upset_df <- function(gwas, thr = 7, quantile = NA, metadata){
  gwas_ok <- floor(gwas$ncol / 3)
  ind_p <- (1:(gwas_ok))*3
  colnames_fbm <- metadata$phe
  if(gwas$ncol < (sum(gwas_ok)*3 + 1))  {
  gwas$add_columns(ncol_add = 1)
    }  # add a column for the threshold score if there isn't one already
  thr_df <- big_apply(gwas, a.FUN = function(X, ind) rowMaxs(as.matrix(gwas[, ind])), 
                      ind = ind_p, a.combine = 'c', block.size = 100)
  gwas[,(sum(gwas_ok)*3 + 1)] <- thr_df
  thr_df <- which(thr_df > thr)
  thr_df <- big_copy(gwas, ind.row = thr_df, ind.col = ind_p)

  for (j in seq_along(1:thr_df$ncol)) {  # standardize one gwas at a time.
  thr_df[, j] <- big_apply(thr_df, a.FUN = thresholdFBM, ind = j, thr = thr,
                              a.combine = 'plus')
  }
  
  thr_df1 <- thr_df[1:thr_df$nrow, 1:thr_df$ncol]
  colnames(thr_df1) <- colnames_fbm
  thr_df1 <- as_tibble(thr_df1)
  return(thr_df1)
}

```

```{r}
if (!dir.exists(here("analysis", "gwas", "permutations"))) {
  dir.create(here("analysis", "gwas", "permutations"))
}

thr_fbm <- big_upset_df(m_pan$effects, thr = 6.5, metadata = m_pan$metadata)

upset_plot <- upset(thr_fbm, colnames(thr_fbm), min_size = 0, name = "Significant at -log10p > 6.5", sort_sets=FALSE)
save_plot(filename = here("analysis", "gwas", "permutations",
                          paste0("Upset_plot_",
                                 "four_panicle_permutation_types",".png")),
          plot = upset_plot, base_height = 12)

thr_fbm7 <- big_upset_df(m_pan$effects, thr = 7, metadata = m_pan$metadata)

upset_plot7 <- upset(thr_fbm7, colnames(thr_fbm7), min_size = 0, name = "Significant at -log10p > 7", sort_sets=FALSE)
save_plot(filename = here("analysis", "gwas", "permutations",
                          paste0("Upset_plot_thr_7_",
                                 "four_panicle_permutation_types",".png")),
          plot = upset_plot7, base_height = 12)

thr_fbmb <- big_upset_df(m_pan$effects, thr = 8.573, metadata = m_pan$metadata)

upset_plotb <- upset(thr_fbmb, colnames(thr_fbmb), min_size = 0, name = "Significant at -log10p > Bonferroni", sort_sets = FALSE)
save_plot(filename = here("analysis", "gwas", "permutations",
                          paste0("Upset_plot_thr_Bonferroni_",
                                 "four_panicle_permutation_types",".png")),
          plot = upset_plotb, base_height = 12)
-log10(18704908/0.05)
```
this one plot honestly answers any upset plot questions for this guy. Basically no overlaps between any permuted phenotype. Ok.


## --------
## Run dive_phe2mash on list of panicle phe
```{r}
permuted_phe <- readRDS(here("data/permuted_phenotypes_geno_subset.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)
permuted_phe_subpop_ecotype <- 
  readRDS(here("data/permuted_phenotypes_within_subpop_and_ecotype_geno_subset.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)
permuted_phe_subpop <- 
  readRDS(here("data/permuted_phenotypes_within_subpops_geno_subset.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)

(phe_gwas_list <- list(permuted_phe = permuted_phe,
                      permuted_phe_by_subpop = permuted_phe_subpop,
                      permuted_phe_by_subpop_and_ecotype = permuted_phe_subpop_ecotype))

```


```{r}
subpop_v2 <- list(Gulf_and_Midwest = c("Gulf", "Midwest"), Midwest = "Midwest", 
                  Gulf = "Gulf", Atlantic = "Atlantic", 
                  Atlantic_and_Midwest = c("Atlantic", "Midwest"),
                  Atlantic_and_Gulf = c("Atlantic", "Gulf"), 
                  All = c("Atlantic", "Gulf", "Midwest", "4X"))
k = 7
suffix_svd <- paste0("big_three_garden_SNPs", names(subpop_v2)[k])
pavir_svd <- readRDS(file = here("analysis", "mash", paste0(suffix_svd,
                                                        "_svd.rds")))
  
bigsnp_inputs <- read_delim(here("analysis", 
                                   "bigsnp_inputs.txt"),  delim = " ", 
                                   col_names = "SNPfiles") 

```

```{r}
if (!dir.exists(here("analysis", "mash", "permutations"))) {
  dir.create(here("analysis", "mash", "permutations"))
}

for (i in seq_along(phe_gwas_list)) {
  pavir_pan <- phe_gwas_list[[i]]
  suffix = paste0(names(phe_gwas_list)[i])
  pavir_snp <- snp_attach(bigsnp_inputs$SNPfiles[k])
  
  m_pan <- dive_phe2effects(df = pavir_pan, snp = pavir_snp, svd = pavir_svd,
                       type = "linear", suffix = suffix, ncores = 11, 
                       outputdir = here("analysis", "mash", "replicates"))
}

for (i in seq_along(phe_gwas_list)) {
  suffix = paste0(names(phe_gwas_list)[i])
  pavir_gwas <- big_attach(here("analysis", "mash", "replicates", 
                                paste0("gwas_effects_", suffix, ".rds")))
  pavir_meta <- read_csv(here("analysis", "mash", "replicates", 
                                paste0("gwas_effects_", suffix, 
                                       "_associated_metadata.csv")))
  pavir_snp <- snp_attach(bigsnp_inputs$SNPfiles[k])
  
  m_pan <- dive_effects2mash(effects = pavir_gwas, metadata = pavir_meta, 
                             snp = pavir_snp, suffix = suffix, thr.m = "max",
                             outputdir = here("analysis", "mash", 
                                              "permutations"),
                             num.strong = 5000, num.random = 100000)
  saveRDS(m_pan, file = here("analysis", "mash", "permutations", 
                             paste0("Full_mash_model_",
                                    "5000", "_SNPs_U_ed_and_",
                                    "100000", "_SNPs_mash_fit_",
                                                    suffix, ".rds")))
  mash_plot_covar(m_cpan, savceoutput = TRUE, suffix = suffix)

  manhattan <- mash_plot_manhattan_by_condition(m_pan, snp = pavir_snp, 
                                                saveoutput = TRUE)
#  mash_clumps <- snp_clumping(pavir_snp$genotypes, 
#                              infos.chr = markers$CHRN, thr.r2 = 0.2, 
#                              infos.pos = markers$POS, 
#                              S = manhattan$ggman_df$log10BayesFactor)
#  mash_df_clumped <- manhattan$ggman_df[mash_clumps,]
#  write_csv(mash_df_clumped, file = here("analysis", "mash", "permutations",
#                                              paste0("Clumped_mash_output_df_",
#                                                     suffix, ".csv")))
}

```

# ---------

# Mash or just GWAS on replicates to look for major conserved hits, probably with upset plots

## Set up data for GWAS
could potentially resample the data too, shuffle which replicate belongs to which, then compare.

```{r}
phe <- readRDS(here("data/panicle_phenotypes_genotype_subset.rds"))
phe <- phe %>%
  mutate(phe_name = case_when(PHE == "PAN_LEN" ~ "PL",
                              PHE == "PRIM_BN" ~ "PBN",
                              PHE == "SEC_BN" ~ "SBN"))
phe$ECOTYPE_NNET <- factor(phe$ECOTYPE_NNET, levels = c("Upland", "Coastal", "Lowland", "Unknown"))
phe$SITE <- factor(phe$SITE, levels = c("PKLE", "CLMB", "KBSM"))
phe$phe_name <- factor(phe$phe_name, levels = c("PL", "PBN", "SBN"))

pheA <- phe %>%
  filter(REP == "A") %>%
  select(-BLACKWELL, -manu_site, -PHE, -PLOT_GL, -REP, -YEAR) %>%
  pivot_wider(names_from = c("phe_name", "SITE"), values_from = MEAS, 
              values_fn = mean)
pheB <- phe %>%
  filter(REP == "B") %>%
  select(-BLACKWELL, -manu_site, -PHE, -PLOT_GL, -REP, -YEAR) %>%
  pivot_wider(names_from = c("phe_name", "SITE"), values_from = MEAS, 
              values_fn = mean)
pheC <- phe %>%
  filter(REP == "C") %>%
  select(-BLACKWELL, -manu_site, -PHE, -PLOT_GL, -REP, -YEAR) %>%
  pivot_wider(names_from = c("phe_name", "SITE"), values_from = MEAS, 
              values_fn = mean)

phe_notperm <- phe %>%
  filter((PLANT_ID == "J636.B" & SITE == "KBSM"))

pheD <- phe %>%
  filter(!(PLANT_ID == "J636.B" & SITE == "KBSM")) %>%
  select(-BLACKWELL, -manu_site, -PHE, -PLOT_GL, -YEAR) %>%
  group_by(PLANT_ID, phe_name, SITE) %>%
  mutate(MEAS_perm = sample(MEAS)) %>%
  filter(REP == "A") %>%
  full_join(phe_notperm) %>%
  select(-BLACKWELL, -manu_site, -PHE, -PLOT_GL, -REP, -YEAR, -MEAS) %>%
  pivot_wider(names_from = c("phe_name", "SITE"), values_from = MEAS_perm, 
              values_fn = mean)
pheE <- phe %>%
  filter(!(PLANT_ID == "J636.B" & SITE == "KBSM")) %>%
  select(-BLACKWELL, -manu_site, -PHE, -PLOT_GL, -YEAR) %>%
  group_by(PLANT_ID, phe_name, SITE) %>%
  mutate(MEAS_perm = sample(MEAS)) %>%
  filter(REP == "C") %>%
  full_join(phe_notperm) %>%
  select(-BLACKWELL, -manu_site, -PHE, -PLOT_GL, -REP, -YEAR, -MEAS) %>%
  pivot_wider(names_from = c("phe_name", "SITE"), values_from = MEAS_perm, 
              values_fn = mean)


saveRDS(pheA, file = here("data/raw_phenotypes_replicate_A_GWAS.rds"))
saveRDS(pheB, file = here("data/raw_phenotypes_replicate_B_GWAS.rds"))
saveRDS(pheC, file = here("data/raw_phenotypes_replicate_C_GWAS.rds"))
saveRDS(pheD, file = here("data/raw_phenotypes_replicate_D_GWAS.rds"))
saveRDS(pheE, file = here("data/raw_phenotypes_replicate_E_GWAS.rds"))
```

## Set up GWAS & mash

```{r}
phe_repA <- readRDS(here("data/raw_phenotypes_replicate_A_GWAS.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)
phe_repB <- readRDS(here("data/raw_phenotypes_replicate_B_GWAS.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)
phe_repC <- readRDS(here("data/raw_phenotypes_replicate_C_GWAS.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)
phe_repD <- readRDS(here("data/raw_phenotypes_replicate_D_GWAS.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)
phe_repE <- readRDS(here("data/raw_phenotypes_replicate_E_GWAS.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET) %>%
  rename(sample.ID = .data$PLANT_ID)

phe_rep_all <- phe_repA %>%
  full_join(phe_repB, by = "sample.ID") %>%
  full_join(phe_repC, by = "sample.ID") %>%
  full_join(phe_repD, by = "sample.ID") %>%
  full_join(phe_repE, by = "sample.ID")

colnames(phe_rep_all) <- str_replace_all(colnames(phe_rep_all), ".x.x", "_C")
colnames(phe_rep_all) <- str_replace_all(colnames(phe_rep_all), ".y.y", "_D")
colnames(phe_rep_all) <- str_replace_all(colnames(phe_rep_all), ".x", "_A")
colnames(phe_rep_all) <- str_replace_all(colnames(phe_rep_all), ".y", "_B")
colnames(phe_rep_all)[38:46] <- paste0(colnames(phe_rep_all)[38:46], "_E")

phe_rep_all <- phe_rep_all %>%
  select(sample.ID, starts_with("PL_PKLE"), starts_with("PL_CLMB"), 
         starts_with("PL_KBSM"), starts_with("PBN_PKLE"),
         starts_with("PBN_CLMB"), starts_with("PBN_KBSM"), 
         starts_with("SBN_PKLE"),
         starts_with("SBN_CLMB"), starts_with("SBN_KBSM")
         )
```

```{r}
subpop_v2 <- list(Gulf_and_Midwest = c("Gulf", "Midwest"), Midwest = "Midwest", 
                  Gulf = "Gulf", Atlantic = "Atlantic", 
                  Atlantic_and_Midwest = c("Atlantic", "Midwest"),
                  Atlantic_and_Gulf = c("Atlantic", "Gulf"), 
                  All = c("Atlantic", "Gulf", "Midwest", "4X"))
k = 7
suffix_svd <- paste0("big_three_garden_SNPs", names(subpop_v2)[k])
pavir_svd <- readRDS(file = here("analysis", "mash", paste0(suffix_svd,
                                                        "_svd.rds")))
  
bigsnp_inputs <- read_delim(here("analysis", 
                                   "bigsnp_inputs.txt"),  delim = " ", 
                                   col_names = "SNPfiles") 
```


## Run dive_phe2mash
```{r}
if (!dir.exists(here("analysis", "mash", "replicates"))) {
  dir.create(here("analysis", "mash", "replicates"))
}

suffix = "replicated_panicle_phe"
pavir_snp <- snp_attach(bigsnp_inputs$SNPfiles[k])

m_pan <- dive_phe2effects(df = phe_rep_all, snp = pavir_snp, svd = pavir_svd,
                       type = "linear", suffix = suffix, ncores = 8, 
                       outputdir = here("analysis", "mash", "replicates"))

```

```{r}
m_pan2 <- dive_effects2mash(effects = m_pan$effects, snp = pavir_snp, 
                            metadata = m_pan$metadata, 
                            suffix = "replicated_panicle_phe_5Kstrong_100Krandom",
                            outputdir = here("analysis", "mash",
                                             "replicates"), ncores = 12,
                            num.strong = 5000, num.random = 100000)

  saveRDS(m_pan2, file = here("analysis", "mash", "replicates", 
                             paste0("Full_mash_model_",
                                    "5000", "_SNPs_U_ed_and_",
                                    "100000", "_SNPs_mash_fit_",
                                                    suffix, ".rds")))
  mash_plot_covar(m_pan2, saveoutput = TRUE, suffix = suffix)

  manhattan <- mash_plot_manhattan_by_condition(m_pan2, snp = pavir_snp, 
                                                saveoutput = TRUE)
```


# ---
# Upset plots of replicates
```{r}

library(ComplexUpset)
library(cowplot)
library(matrixStats)

thresholdFBM <- function(X, ind, thr) { #, quantile = NA) {
  #if(!is.na(quantile)) {
  #thr <- quantile(X[, ind], quantile)
  #}
  case_when(X[, ind] > thr ~ TRUE,
            TRUE ~ FALSE)
}

big_upset_df <- function(gwas, thr = 7, quantile = NA, metadata){
  gwas_ok <- floor(gwas$ncol / 3)
  ind_p <- (1:(gwas_ok))*3
  colnames_fbm <- metadata$phe
  if(gwas$ncol < (sum(gwas_ok)*3 + 1))  {
  gwas$add_columns(ncol_add = 1)
    }  # add a column for the threshold score if there isn't one already
  thr_df <- big_apply(gwas, a.FUN = function(X, ind) rowMaxs(as.matrix(gwas[, ind])), 
                      ind = ind_p, a.combine = 'c', block.size = 100)
  gwas[,(sum(gwas_ok)*3 + 1)] <- thr_df
  thr_df <- which(thr_df > thr)
  thr_df <- big_copy(gwas, ind.row = thr_df, ind.col = ind_p)

  for (j in seq_along(1:thr_df$ncol)) {  # standardize one gwas at a time.
  thr_df[, j] <- big_apply(thr_df, a.FUN = thresholdFBM, ind = j, thr = thr,
                              a.combine = 'plus')
  }
  
  thr_df1 <- thr_df[1:thr_df$nrow, 1:thr_df$ncol]
  colnames(thr_df1) <- colnames_fbm
  thr_df1 <- as_tibble(thr_df1)
  return(thr_df1)
}

```

```{r}
thr_fbm <- big_upset_df(m_pan$effects, thr = 6.5, metadata = m_pan$metadata)

upset_plot <- upset(thr_df, colnames(thr_df), min_size = 25, name = "Significant at -log10p > 6.5")
save_plot(filename = here("analysis", "mash", "replicates",
                          paste0("Upset_plot_", "5_panicle_replicates",".png")),
          plot = upset_plot, base_height = 12)

upset_plot <- upset(thr_df, colnames(thr_df)[1:15], min_size = 5, name = "Significant at -log10p > 6.5")
save_plot(filename = here("analysis", "mash", "replicates",
                          paste0("Upset_plot_", "5_panicle_replicates_", 
                                 "PL", ".png")),
          plot = upset_plot, base_height = 12)

upset_plot <- upset(thr_df, colnames(thr_df)[16:30], min_size = 5, name = "Significant at -log10p > 6.5")
save_plot(filename = here("analysis", "mash", "replicates",
                          paste0("Upset_plot_", "5_panicle_replicates_", 
                                 "PBN", ".png")),
          plot = upset_plot, base_height = 12)

upset_plot <- upset(thr_df, colnames(thr_df)[31:45], min_size = 5, name = "Significant at -log10p > 6.5")
save_plot(filename = here("analysis", "mash", "replicates",
                          paste0("Upset_plot_", "5_panicle_replicates_", 
                                 "SBN", ".png")),
          plot = upset_plot, base_height = 12)


```

These really don't overlap. May need to bin by 20kb region again and try and make Upset plots of those overlaps instead. Ugh.