---
title: "Permuted phenotypes 2"
author: "Alice MacQueen"
date: "11/23/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(bigsnpr)
# library(switchgrassGWAS)

```

# New SNP files

Tom asked if we get similar GWAS when using a SNP file with a higher MAF, say 10%.

Also, I want to get mash to run on these GWAS results, maybe using switchgrassGWAS, maybe using snpdiver. Not sure what bugfixes will be required to achieve this yet.

```{r load data}
snp <- snp_attach("~/Github/pvdiv-genome/big_three/Pvirgatum_V5_GWAS_630g_33M_big_three.rds")
plants <- snp$fam$sample.ID

metadata <- readRDS(here("data/metadata.rds"))
phe_df <- readRDS(here("data/all_permuted_phenotypes_geno_subset.rds"))
plink <- download_plink()
```

```{r pop subset function}

pvdiv_pop_subset_bigsnp <- function(snp, subset, selector, maf = NA){
  if (is.na(maf)) {
    mac <- round(25/length(selector), digits = 2) # what should the MAF be?
    if(mac > 0.05){
      mac <- 0.05
    }
  } else {
    mac <- maf
  }
  
  subset_name <- paste0("Pvirgatum_V5_GWAS_", length(selector), "g_", subset, 
                        "_subset_maf_", mac)
  snpfile_subset <- subset(snp, ind.row = selector)
  snp_subset <- snp_attach(snpfile_subset)
  snp_writeBed(snp_subset, bedfile = paste0(subset_name, ".bed"))
  
  bedfile <- snp_plinkQC(plink, prefix.in = subset_name,
                     maf = mac, geno = 0.1, hwe = 1e-50, 
                     prefix.out = paste0(subset_name, "_maf", mac),
                     extra.options = "--allow-no-sex --allow-extra-chr --chr Chr01K Chr01N Chr02K Chr02N Chr03K Chr03N Chr04K Chr04N Chr05K Chr05N Chr06K Chr06N Chr07K Chr07N Chr08K Chr08N Chr09K Chr09N")
  rdsfile <- snp_readBed(bedfile)
}

```

```{r make selector}
panicle_381 <- which(plants %in% phe_df$PLANT_ID)

```

```{r make subset}
pvdiv_pop_subset_bigsnp(snp = snp, subset = "PanicleData", 
                        selector = panicle_381, maf = 0.05)
# 381 genotypes and 12.7M SNPs
pvdiv_pop_subset_bigsnp(snp = snp, subset = "PanicleData", 
                        selector = panicle_381, maf = 0.10)
# 381 genotypes and 7.4M SNPs
```

```{r}
381*.05
```
# --------------
# Analysis of these two new subsets


## GWAS on real and permuted data
```{r prepare phenotypes}
(all_permuted_phe <- readRDS(here("data/all_permuted_phenotypes_geno_subset.rds")) %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET))

prefix <- "pvdiv_panicles_2019_BLUPs_PLANT_ID_geno_subset_"
(phe_gwas <- read_csv(here("data", 
                          paste0(prefix, "BLUP_phenotypes",
                                              ".csv"))) %>%
  pivot_longer(2:13, names_to = "PHE_SITE", values_to = "MEAS") %>%
  mutate(PHE_SITE = str_replace_all(PHE_SITE, "PAN_LEN", "PL"),
         PHE_SITE = str_replace_all(PHE_SITE, "SEC_BN", "SBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "PRIM_BN", "PBN"),
         PHE_SITE = str_replace_all(PHE_SITE, "TX2", "PKLE"),
         PHE_SITE = str_replace_all(PHE_SITE, "MO", "CLMB"),
         PHE_SITE = str_replace_all(PHE_SITE, "MI", "KBSM")) %>%
  filter(!grepl("SEC_LN", PHE_SITE)) %>%
  pivot_wider(names_from = PHE_SITE, values_from = MEAS, names_prefix = "BLUP_"))

(all_raw_reps <- readRDS(file = here("data/all_replicates_raw_phenotypes_geno_subset.rds")) %>%
  rename(PLANT_ID = sample.ID))

permuted_phe2 <- readRDS(here("data/permuted_phenotypes_geno_subset.rds"))
permuted_phe2 <- permuted_phe2 %>%
  ungroup() %>%
  select(-SUBPOP, -ECOTYPE_NNET)

all_phe <- phe_gwas %>%
  full_join(permuted_phe2) %>%
  full_join(all_permuted_phe) %>%
  full_join(all_raw_reps) %>%
  select(PLANT_ID, contains("PL_PKLE"), contains("PL_CLMB"), 
         contains("PL_KBSM"), contains("PBN_PKLE"),
         contains("PBN_CLMB"), contains("PBN_KBSM"), 
         contains("SBN_PKLE"),
         contains("SBN_CLMB"), contains("SBN_KBSM"), everything()
         )
saveRDS(all_phe, file = here("data/all_blup_raw_and_permuted_panicle_phe_geno_subset.rds"))

all_phe %>%
  select(PLANT_ID, BLUP_PL_PKLE, BLUP_PBN_KBSM, BLUP_SBN_PKLE) %>%
  saveRDS(here("data/test_phe_switchgrassGWAS.rds"))
```

```{r load files}
snp10 <- snp_attach("~/Github/PanicleData/9-DiversityPanel/data/Pvirgatum_V5_GWAS_381g_PanicleData_subset_maf_0.1.rds")
snp05 <- snp_attach("~/Github/PanicleData/9-DiversityPanel/data/Pvirgatum_V5_GWAS_381g_PanicleData_subset_maf_0.05.rds")

testphe <- readRDS("~/Github/PanicleData/9-DiversityPanel/data/test_phe_switchgrassGWAS.rds")
all_phe <- readRDS(file = here("data/all_blup_raw_and_permuted_panicle_phe_geno_subset.rds"))
covar10 <- readRDS("~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe/SVD_381g_7.4M_SNPs_15PCs.rds")
plot(covar10)
plot(covar10, type = "scores")
```



```{r ten percent maf GWAS}
if (!dir.exists(here("analysis", "gwas", "MAF10"))) {
  dir.create(here("analysis", "gwas", "MAF10"))
}
pvdiv_standard_gwas(snp = snp10, df = all_phe, covar = covar10, 
                            type = "linear", 
                            outputdir = here("analysis", "gwas", "MAF10"), 
                            savegwas = TRUE, savetype = 'fbm', 
                            suffix = "all_panicle_phe_MAF_10per")
```


```{r five percent maf GWAS}
if (!dir.exists(here("analysis", "gwas", "MAF05"))) {
  dir.create(here("analysis", "gwas", "MAF05"))
}
pvdiv_standard_gwas(snp = snp05, df = all_phe, 
                            type = "linear", 
                            outputdir = here("analysis", "gwas", "MAF05"), 
                            savegwas = TRUE, savetype = 'fbm', 
                            suffix = "all_panicle_phe_MAF_5per")
```


## GWAS on BLUP & more noisy replicate data
```{r}
library(tidyverse)
library(bigsnpr)
library(ComplexUpset)
library(cowplot)

effects <- big_attach("~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe/gwas_effects_test_phe.rds")

metadata <- read_csv("~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe/gwas_effects_test_phe_associated_metadata.csv")

```

```{r}
upset1 <- get_fbm_upset_df(effects = gwas, metadata = gwas_metadata)
upsetp <- upset(upset1, colnames(upset1)[-(1:2)])
save_plot(filename = "./Upset_test.png", plot = upsetp)

```


## mash on real and permuted data

```{r}
mtest <- dive_effects2mash(effects = effects, snp = snp10, metadata = metadata, 
                  suffix = "test_mash", 
                  outputdir = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe")


snp10_sub <- subset(snp10, ind.col = c(1, 1000000, 2000000, 3000000, 4000000, 5000000, 6000000, 7000000))
snp_10s <- snp_attach(snp10_sub)
effects_sub <- big_copy(effects, ind.row = c(1,1000000,2000000,3000000,4000000,5000000,6000000,7000000))
mtest <- dive_effects2mash(effects = effects_sub, snp = snp_10s, metadata = metadata, 
                  suffix = "test_mash", 
                  outputdir = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe")
        

mtest <- dive_effects2mash(effects = effects_sub, snp = snp_10s, metadata = metadata, phe = c(1,2),
                  suffix = "test_mash", 
                  outputdir = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe")
             
```

# test dive_effects2mash on just some columns of an effects fbm

```{r}
subsample <- c(1:7400)*1000

snp10 <- snp_attach("~/Github/PanicleData/9-DiversityPanel/data/Pvirgatum_V5_GWAS_381g_PanicleData_subset_maf_0.1.rds")
effects_all <- big_attach("~/Github/PanicleData/9-DiversityPanel/analysis/gwas/MAF10/gwas_effects_all_panicle_phe_MAF_10per.rds")

metadata <- read_csv("~/Github/PanicleData/9-DiversityPanel/analysis/gwas/MAF10/gwas_effects_all_panicle_phe_MAF_10per_associated_metadata.csv")

snp10_sub <- subset(snp10, ind.col = subsample)
snp_10s <- snp_attach(snp10_sub)
effects_sub <- big_copy(effects, ind.row = subsample)

mtest <- dive_effects2mash(effects = effects_sub, snp = snp_10s, metadata = metadata, phe = c(1,7:11), suffix = "test_mash_3", 
                  outputdir = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe")

mtest <- dive_effects2mash(effects = effects_sub, snp = snp_10s, 
                           metadata = metadata, 
                  suffix = "test_mash_2", 
                  outputdir = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe")

mtest4 <- dive_effects2mash(effects = effects_all, snp = snp10, 
                           metadata = metadata, 
                  suffix = "test_mash_4", 
                  outputdir = "~/Github/PanicleData/9-DiversityPanel/analysis/gwas/testphe")
         

```


```{r}
effects10 <- big_attach(here("analysis/gwas/MAF10/gwas_effects_all_panicle_phe_MAF_10per.rds"))
metadata10 <- read_csv(here("analysis/gwas/MAF10/gwas_effects_all_panicle_phe_MAF_10per_associated_metadata.csv"))
```

```{r}
library(ComplexUpset)
upset_10_thr_8 <- pvdiv_fbm_upset_df(effects = effects10, metadata = metadata10,
                                     snp = snp10, thr = 8)
upset_10_thr_7 <- pvdiv_fbm_upset_df(effects = effects10, metadata = metadata10,
                                     snp = snp10, thr = 7)
upset_10_thr_6 <- pvdiv_fbm_upset_df(effects = effects10, metadata = metadata10,
                                     snp = snp10, thr = 6)
upset_10_thr_5 <- pvdiv_fbm_upset_df(effects = effects10, metadata = metadata10,
                                     snp = snp10, thr = 5)

upset_plot_10thr6 <- upset(upset_10_thr_6, intersect = colnames(upset_10_thr_6)[(3:35)], name = "-log10p > 6", min_size = 10)
upset_plot_10thr6

upset1 <- upset_10_thr_5 %>%
  select(contains("PERM"), contains("BLUP")) %>%
  select(!contains("IND_PERM"))
upsetp1 <- upset(upset1, intersect = colnames(upset1), name = "-log10p > 5",
                 min_size = 5, sort_sets=FALSE)
upsetp1

upset1 <- upset_10_thr_5 #%>%
  #select(contains("IND_PERM"), contains("BLUP")) 
upsetp1 <- upset(upset1, intersect = colnames(upset1), name = "-log10p > 5",
                 min_size = 5, sort_sets=FALSE)
upsetp1

upset1 <- upset_10_thr_5 %>%
  select(ends_with("_A"),ends_with("_B"),ends_with("_C"),ends_with("_D"),ends_with("_E"), contains("BLUP")) 
upsetp1 <- upset(upset1, intersect = colnames(upset1), name = "-log10p > 5",
                 min_size = 5, sort_sets=FALSE)
upsetp1
```


# Mash tests in testphe

```{r}
m1 <- readRDS(here("analysis/gwas/testphe/mash_test_2.rds"))

mash_plot_covar(m1, saveoutput = TRUE)
tpca <- mash_plot_Ulist(m1, range = 6)
length(get_significant_results(m1))
mash_plot_effects(m1, n = 1)
mash_plot_effects(m1, n = 2)
mash_plot_effects(m1, n = 6333)
mash_plot_sig_by_condition(m1, which(grepl("ALL_PERM", switchgrassGWAS:::get_colnames(m1)) == TRUE))

snpdiver::mash_plot_Ulist(m1, range = 6, labels = FALSE)

tpca$ED_tPCA_df %>%
  filter(grepl("BLUP", rowU)) %>%
  filter(abs(covar) > 0.05) %>%
  filter(grepl("PERM", colU))
tpca$ED_tPCA_df %>%
  filter(grepl("BLUP", rowU)) %>%
  filter(abs(covar) > 0.05) %>%
  arrange(colU)

tpca$ED_tPCA_df %>%
  filter(grepl("ALL_PERM", rowU)) %>%
  filter(abs(covar) > 0.05)
tpca$ED_tPCA_df %>%
  filter(grepl("SUB_PERM", rowU)) %>%
  filter(abs(covar) > 0.05)
```

